src/mlpack/core/data/split_data.hpp

```diff
@@ -61,20 +64,21 @@ void Split(const arma::Mat<T>& input,
   trainLabel.set_size(trainSize);
   testLabel.set_size(testSize);
 
-  const arma::Col<size_t> order =
-      arma::shuffle(arma::linspace<arma::Col<size_t>>(0, input.n_cols - 1,
-                                                      input.n_cols));
-
-  for (size_t i = 0; i != trainSize; ++i)
+  if (shuffleData)
   {
-    trainData.col(i) = input.col(order[i]);
-    trainLabel(i) = inputLabel(order[i]);
+    arma::uvec order = arma::shuffle(arma::linspace<arma::uvec>(
+        0, input.n_cols - 1, input.n_cols));
+    trainData = input.cols(order.subvec(0, trainSize - 1));
+    trainLabel = inputLabel.cols(order.subvec(0, trainSize - 1));
+    testData = input.cols(order.subvec(trainSize, input.n_cols - 1));
+    testLabel = inputLabel.cols(order.subvec(trainSize, input.n_cols - 1));
```

Comment:
To fix the failing tests, we have to account for when `ratio` is `0` or `1`.  In these edge cases, `trainSize` will be greater than `input.n_cols - 1` (causing an error), or `trainSize` will be 0.  So, if you encase these blocks in, e.g., `if (trainSize > 0) ...` and `if (trainSize < input.n_cols)` ... then I think this should fix the issue. :+1:

Reply:
How are you so good ?

How did you come up with it ?

Reply:
Oh, I didn't do anything special. :) I checked out the branch and ran the test; when I saw it was failing, I started `gdb` and then had it break on any exceptions with `catch throw`.  Then, I ran the test, and when it threw the exception, I just looked at where the code was in `Split()`.  Quick inspection of `trainSize` and `input.n_cols` showed that `trainSize` was larger than `input.n_cols`, so then I realized what was happening. :)

Reply:
I added a test , so that next time someone changed implementation they take care of these edge cases